mava/systems/q_learning/sebulba/rec_iql.py

```
@@ -0,0 +1,782 @@
+        rollout_queue: Queue for sending collected trajectories to the learner.
+        params_source: Provides the latest network parameters from the learner.
+        q_net: The Q-network.
+        actor_device: Index of the actor device to use for rollout.
+        seeds: Seeds for environment initialization.
+        thread_lifetime: Controls the thread's lifecycle.
+        actor_id: Unique identifier for the actor.
+    """
+    name = threading.current_thread().name
+    print(f"{Fore.BLUE}{Style.BRIGHT}Thread {name} started{Style.RESET_ALL}")
+    num_agents = config.system.num_agents
+    move_to_device = lambda x: jax.device_put(x, device=actor_device)
+
+    @jax.jit
+    def select_eps_greedy_action(
+        params: FrozenDict,
+        hidden_state: jax.Array,
+        obs: Observation,
+        term_or_trunc: Array,
+        key: chex.PRNGKey,
+        t: int,
+    ) -> Tuple[Array, Array, int]:
+        """Selects an action epsilon-greedily.
+
+        Args:
+            params: Network parameters.
+            hidden_state: Current RNN hidden state.
+            obs: Observation from the environment.
+            term_or_trunc: Termination or truncation flag.
+            key: PRNG key for sampling.
+            t: Current timestep (used for epsilon decay).
+
+        Returns:
+            Tuple containing the chosen action, next hidden state, and updated timestep.
+        """
+
+        eps = jnp.maximum(
+            config.system.eps_min, 1 - (t / config.system.eps_decay) * (1 - config.system.eps_min)
+        )
+
+        obs = tree.map(lambda x: x[jnp.newaxis, ...], obs)
+        term_or_trunc = tree.map(lambda x: x[jnp.newaxis, ...], term_or_trunc)
+
+        next_hidden_state, eps_greedy_dist = q_net.apply(
+            params, hidden_state, (obs, term_or_trunc), eps
+        )
+
+        action = eps_greedy_dist.sample(seed=key)
+        action = action[0, ...]  # (1, B, A) -> (B, A)
```

Bit safer as this will silently error if actions 0 dim is ever larger than 1

```suggestion
        action = eps_greedy_dist.sample(seed=key).squeeze(0)  # (B, A)
```